test/vfd_swmr_group_writer.c

```diff
@@ -2981,58 +3304,2076 @@ verify_group(state_t *s, unsigned int which)
-     * as the second parameter of H5Pset_libver_bound() that is called by
-     * vfd_swmr_create_fapl. Otherwise, the latest file format(H5F_LIBVER_LATEST)
-     * should be used as the second parameter of H5Pset_libver_bound().
-     * Also pass the use_vfd_swmr, only_meta_page, config to vfd_swmr_create_fapl().*/
-    if ((fapl = vfd_swmr_create_fapl(!s.old_style_grp, s.use_vfd_swmr, true, &config)) < 0) {
+    if (which >= s->nsteps) {
         H5_FAILED();
         AT();
-        printf("vfd_swmr_create_fapl failed\n");
+        printf("Number of created groups is out of bounds\n");
         goto error;
     }
 
-    if ((fcpl = H5Pcreate(H5P_FILE_CREATE)) < 0) {
+    gcpl = H5Pcreate(H5P_GROUP_CREATE);
+    if (gcpl < 0) {
+        H5_FAILED();
+        AT();
+        printf("H5Pcreate failed\n");
+        goto error;
+    }
+
+    if (dense_to_compact) {
+        if (H5Pset_link_phase_change(gcpl, 2, 2) < 0) {
+            H5_FAILED();
+            AT();
+            printf("H5Pset_link_phase_change failed for dense to compact.\n");
+            goto error;
+        }
+    }
+    else {
+        if (H5Pset_link_phase_change(gcpl, 1, 1) < 0) {
+            H5_FAILED();
+            AT();
+            printf("H5Pset_attr_phase_change failed for compact to dense.\n");
+            goto error;
+        }
+    }
+
+    esnprintf(name, sizeof(name), "/group-%u", which);
+    if ((g = H5Gcreate2(s->file, name, H5P_DEFAULT, gcpl, H5P_DEFAULT)) < 0) {
+        H5_FAILED();
+        AT();
+        printf("H5Gcreate2 failed\n");
+        goto error;
+    }
+
+    if (H5Gget_info(g, &group_info) < 0) {
+        H5_FAILED();
+        AT();
+        printf("H5Gget_info failed\n");
+        goto error;
+    }
+
+    /* The storage type should always be compact when a group is created. */
+    if (group_info.storage_type != H5G_STORAGE_TYPE_COMPACT) {
+        H5_FAILED();
+        AT();
+        printf("New-style group link storage test:. \n");
+        printf("    still be compact after group creation. \n");
+        goto error;
+    }
+
+    if (H5Pclose(gcpl) < 0) {
+        H5_FAILED();
+        AT();
+        printf("H5Pclose failed\n");
+        goto error;
+    }
+
+    /* If a grp_op_test is turned on and named pipes are used,
+     * the writer should send and receive messages after the group creation.
+     * Writer sends a message to reader: a group is successfully created.
+     * then wait for the reader to verify and send an acknowledgement message back.*/
+    if (s->use_named_pipes && s->grp_op_test == true) {
+        dbgf(2, "Writer: ready to send the message: %d\n", s->np_notify + 1);
+        if (np_wr_send_receive(s) == false) {
+            H5_FAILED();
+            AT();
+            /* Note: This is (mostly) because the verification failure message
+             *       from the reader. So don't send the error message back to
+             *       the reader. Just stop the test. */
+            goto error2;
```

Comment:
I don't quite understand the "error2" and "error" situation.

Reply:
For the test case that goes to the label "error",  the error occurs on the writer side, so before it exits from the testing,  the writer needs to send  an error message to the reader, signal the reader that  "the test on my side" is wrong, please stop your test. Otherwise, the reader may still wait for a message from the writer. The test on the reader side will hang.
For the test case  that goes to the label "error2",  the error actually occurs on the reader side, the writer receives an error message from the reader to let the test stop, so the writer doesn't need to send an error message to the reader again. It just exits the current test.